File: cpp/63.Unique Paths II.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {

    	/*
    	*	path[i][j] = path[i-1][j] + path[i][j-1]
    	*/

        int path[obstacleGrid.size()][obstacleGrid[0].size()];

        for(int i = 0; i < obstacleGrid.size(); ++i)
			if(!obstacleGrid[i][0])	
        		path[i][0] = i >= 1 ? path[i-1][0] : 1;
        	else
        		path[i][0] = 0;

        for(int j = 0; j < obstacleGrid[0].size(); ++j)
			if(!obstacleGrid[0][j])
        		path[0][j] = j >= 1 ? path[0][j-1] : 1;
        	else
        		path[0][j] = 0;

        for(int i = 1; i < obstacleGrid.size(); ++i)
        	for(int j = 1; j < obstacleGrid[0].size(); ++j)
    			if(!obstacleGrid[i][j])
        			path[i][j] = path[i-1][j] + path[i][j-1];
        		else
        			path[i][j] = 0;

    	return path[obstacleGrid.size() - 1][obstacleGrid[0].size() - 1];
    }
};

File: go/63.Unique_Paths_II.go
// DP
func uniquePathsWithObstacles(obstacleGrid [][]int) int {
	m := len(obstacleGrid)
	n := len(obstacleGrid[0])
	var dp [][]int
	for i := 0; i < m; i++ {
		dp = append(dp, make([]int, n))
	}
	if obstacleGrid[0][0] == 1 {
		return 0
	}
	dp[0][0] = 1
	for i := 1; i < n; i++ {
		if obstacleGrid[0][i] != 1 && dp[0][i-1] == 1 {
			dp[0][i] = 1
		} else {
			dp[0][i] = 0
		}
	}

	for i := 1; i < m; i++ {
		if obstacleGrid[i][0] != 1 && dp[i-1][0] == 1 {
			dp[i][0] = 1
		} else {
			dp[i][0] = 0
		}
	}

	// dp[i][j] = dp[i-1][j] + dp[i][j-1]
	for i := 1; i < m; i++ {
		for j := 1; j < n; j++ {
			if obstacleGrid[i][j] != 1 {
				dp[i][j] = dp[i-1][j] + dp[i][j-1]
			} else {
				dp[i][j] = 0
			}
		}
	}

	return dp[m-1][n-1]
}